compiler/optimizer/OMRSimplifier.hpp

```diff
@@ -121,7 +122,127 @@ class Simplifier : public TR::Optimization
    void prepareToReplaceNode(TR::Node * node);
 
    void anchorOrderDependentNodesInSubtree(TR::Node *node, TR::Node *replacement, TR::TreeTop* anchorTree);
-   TR::Node * unaryCancelOutWithChild(TR::Node * node, TR::Node * firstChild, TR::TreeTop* anchorTree, TR::ILOpCodes opcode, bool anchorChildren=true);
+
+   virtual bool isLegalToUnaryCancel(TR::Node *node, TR::Node *firstChild, TR::ILOpCodes opcode) 
+      { return true; }
+   virtual TR::Node *unaryCancelOutWithChild(TR::Node *node, TR::Node *firstChild, TR::TreeTop *anchorTree, TR::ILOpCodes opcode, bool anchorChildren=true);
+
+   /**
+    * Checks IL tree patterns to see if folding can be done safely.
+    * 
+    * @param[in] node The tree node that we are trying to fold
+    * @param[in] firstChild The first child of node
+    * @return Returns true if node can be folded; false otherwise
+    */
+   virtual bool isLegalToFold(TR::Node *node, TR::Node *firstChild) 
+      { return true; }
+
+   /**
+    * Checks if the bound value is great than or equal to the length value.
+    *
+    * @param[in] boundChild The tree node corresponding to the bound value
+    * @param[in] lengthChild The tree node corresponding to the length value
+    * @return Returns true if bound is great than or equal to length; 
+    * false otherwise
+    */
+   virtual bool isLengthGEBound(TR::Node *boundChild, TR::Node *lengthChild);
```

Comment:
This function name and description is a bit misleading, to me. It's actually asking if the length is known to be >= the bound, isn't it? Even if it returns false, the length may be >= bound. But the way it's named, one could be easily conclude that false means length <  bound. Perhaps name it isLengthDefinitelyGEBound() or something like that?

The actual implementation has nothing to do with lengths and bounds, either. It's just testing whether its first operand is definitely less than the second operand, right? Does it needs to be "tied" to the concept of array bounds checking?